contracts/FractalUsul.sol

```diff
@@ -1,86 +1,469 @@
+        bytes data;
+        Enum.Operation operation;
+    }
+
+    struct Proposal {
+        bool canceled;
+        uint256 timelockPeriod;
+        bytes32[] txHashes;
+        uint256 executionCounter;
+        address strategy; // the module that is allowed to vote on this
+    }
+
+    bytes32 public constant DOMAIN_SEPARATOR_TYPEHASH =
+        0x47e79534a245952e8b16893a336b85a3d9ea9fa8c573f3d803afb92a79469218;
+    bytes32 public constant TRANSACTION_TYPEHASH =
+        0x72e9670a7ee00f5fbf1049b8c38e3f22fab7e9b85029e85cf9412f17fdd5c2ad;
+    uint256 public totalProposalCount; // total number of submitted proposals
+    address internal constant SENTINEL_STRATEGY = address(0x1);
+    mapping(uint256 => Proposal) internal proposals; // Proposals by proposal ID
+    mapping(address => address) internal strategies;
+
+    event ProposalCreated(
+        address strategy,
+        uint256 proposalId,
+        address proposer,
+        Transaction[] transactions,
+        string title,
+        string description,
+        string documentationUrl
+    );
+    event ProposalCanceled(uint256 proposalId);
+    event TransactionExecuted(uint256 proposalId, bytes32 txHash);
+    event TransactionExecutedBatch(uint256 startIndex, uint256 endIndex);
+    event ProposalTimelocked(uint256 proposalId, uint256 endDate);
+    event ProposalExecuted(uint256 id);
+    event UsulSetup(
+        address indexed initiator,
+        address indexed owner,
+        address indexed avatar,
+        address target
+    );
+    event EnabledStrategy(address strategy);
+    event DisabledStrategy(address strategy);
+
+    constructor(
+        address _owner,
+        address _avatar,
+        address _target,
+        address[] memory _strategies
+    ) {
+        bytes memory initParams = abi.encode(
+            _owner,
+            _avatar,
+            _target,
+            _strategies
+        );
+        setUp(initParams);
+    }
+
+    function setUp(bytes memory initParams) public override initializer {
+        (
+            address _owner,
+            address _avatar,
+            address _target,
+            address[] memory _strategies
+        ) = abi.decode(initParams, (address, address, address, address[]));
+        __Ownable_init();
+        avatar = _avatar;
+        target = _target;
+        setupStrategies(_strategies);
+        transferOwnership(_owner);
+        emit UsulSetup(msg.sender, _owner, _avatar, _target);
+    }
+
+    /// @notice Enables a voting strategy that can vote on proposals, only callable by the owner
+    /// @param strategy Address of the strategy to be enabled
+    function enableStrategy(address strategy) public onlyOwner {
+        require(
+            strategy != address(0) && strategy != SENTINEL_STRATEGY,
+            "Invalid strategy"
+        );
+        require(strategies[strategy] == address(0), "Strategy already enabled");
+        strategies[strategy] = strategies[SENTINEL_STRATEGY];
```

Comment:
what is `SENTINEL_STRATEGY` doing exactly? Is that like the default / current strategy? I wonder if there's a better naming we can use for that

Reply:
Yeah so this is an interesting pattern, its basically a linked list that is used to track the strategy contracts. I think of the `SENTINEL_STRATEGY` as the pointer in the linked list. This is the same pattern Gnosis Safe uses for modules, with the same naming you can see here: https://github.com/safe-global/safe-contracts/blob/main/contracts/base/ModuleManager.sol

So this naming does follow Gnosis convention, but I agree its a bit confusing and am alright with renaming it if we'd like to

Reply:
ah, that's cool, yeah I didn't realize it was a known pattern, so let's keep the naming